aws-lc-rs/src/cipher/key.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+use crate::cipher::aes::{encrypt_block_aes, Aes128Key, Aes256Key};
+use crate::cipher::block::Block;
+use crate::cipher::chacha::ChaCha20Key;
+use crate::cipher::AES_128_KEY_LEN;
+use crate::error::Unspecified;
+use aws_lc::{AES_set_decrypt_key, AES_set_encrypt_key, AES_KEY};
+use core::ptr::copy_nonoverlapping;
+use std::mem::{size_of, transmute, MaybeUninit};
+use std::os::raw::c_uint;
+use std::ptr;
+use zeroize::Zeroize;
+
+pub(crate) enum SymmetricCipherKey {
+    Aes128 {
+        raw_key: Aes128Key,
+        enc_key: AES_KEY,
+        dec_key: AES_KEY,
+    },
+    Aes256 {
+        raw_key: Aes256Key,
+        enc_key: AES_KEY,
+        dec_key: AES_KEY,
+    },
+    ChaCha20 {
+        raw_key: ChaCha20Key,
+    },
+}
+
+unsafe impl Send for SymmetricCipherKey {}
+
+// The AES_KEY value is only used as a `*const AES_KEY` in calls to `AES_encrypt`.
+unsafe impl Sync for SymmetricCipherKey {}
+
+impl Drop for SymmetricCipherKey {
+    fn drop(&mut self) {
+        // Aes128Key, Aes256Key and ChaCha20Key implement Drop separately.
+        match self {
+            SymmetricCipherKey::Aes128 {
+                enc_key, dec_key, ..
+            }
+            | SymmetricCipherKey::Aes256 {
+                enc_key, dec_key, ..
+            } => unsafe {
+                #[allow(clippy::transmute_ptr_to_ptr)]
+                let enc_bytes: &mut [u8; size_of::<AES_KEY>()] = transmute(enc_key);
+                enc_bytes.zeroize();
+                #[allow(clippy::transmute_ptr_to_ptr)]
+                let dec_bytes: &mut [u8; size_of::<AES_KEY>()] = transmute(dec_key);
+                dec_bytes.zeroize();
+            },
+            SymmetricCipherKey::ChaCha20 { .. } => {}
+        }
+    }
+}
+
+impl SymmetricCipherKey {
+    pub(crate) fn aes128(key_bytes: &[u8]) -> Result<Self, Unspecified> {
+        if key_bytes.len() != AES_128_KEY_LEN {
+            return Err(Unspecified);
+        }
+
+        unsafe {
```

Comment:
If I'm not mistaken, this is the same for the [AEAD types](https://docs.rs/aws-lc-rs/1.1.0/aws_lc_rs/aead/index.html) as well. I wish it wasn't like this, though, since we have to pay both the latency and memory costs. This is especially wasteful for things like TLS and QUIC where we have separate derived sealer and opener keys for both directions so the other halves never get used.

Reply:
All uni-directional protocols will have to pay this cost. Probably somewhat minor(?), but still.

Reply:
We could lazily initialize these values at the cost of some code complexity and the additional latency of a branch or two in the logic. These are optimizations we should make, but in a separate PR.

On a related note, my [PR here](https://github.com/awslabs/aws-lc-rs/pull/156) eliminates one of the fields of this struct.